x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/SecondaryAuthenticator.java

```diff
@@ -111,13 +111,17 @@ private void authenticate(Consumer<ActionListener<Authentication>> authenticate,
                 listener.onFailure(new ElasticsearchSecurityException("Failed to authenticate secondary user", e));
             })
         );
+
         try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
             logger.trace(
                 "found secondary authentication credentials, placing them in the internal [{}] header for authentication",
                 UsernamePasswordToken.BASIC_AUTH_HEADER
             );
             threadContext.putHeader(UsernamePasswordToken.BASIC_AUTH_HEADER, header);
             authenticate.accept(authenticationListener);
+        } finally {
+            // ensure the restored thread context is sanitized (even if the restored context is not used in practice)
+            threadContext.sanitizeHeaders();
```

Comment:
My preference would be to keep only the sanitize call from inside the `SecurityRestFilter`.
This sanitize doesn't have practical importance, and I think the header lifetime boundary "headers are alive before restHandler.handleRequest" is easier to reason about. In addition, this would also permit using the variables instead of the header string literals.

Reply:
> This sanitize doesn't have practical importance

Correct, however I think it is more correct to either sanitize a restored stashed context or introduce a flushContext() such that we don't stash a context to be restored at all. This line of code does enable the unit test since when the 'ignore' context is closed the original context which still has the un-sanitized headers is restored after all the work is complete. In practice that restored context is not used, but the tests will read the last set context.   

> In addition, this would also permit using the variables instead of the header string literals.

I'm not sure I understand this comment. I was not able to use the constant names for these headers they live in x-pack and thread context is in core. I could pass in a list of names to the sanitize method, that might be a bit cleaner and allow the use of the constant names.

Reply:
> I think it is more correct to either sanitize a restored stashed context or introduce a flushContext() such that we don't stash a context to be restored at all.

I don't follow why that is the case. I still believe that If the sanitize is not necessary then better to not do it.

>  I was not able to use the constant names for these headers they live in x-pack and thread context is in core. I could pass in a list of names to the sanitize method, that might be a bit cleaner and allow the use of the constant names.

I meant the headers can all be referred by variables in `SecurityRestFilter#handleRequest`.

Reply:
My gripe is that the changes overall make it hard to understand the lifetime of the headers (which headers and when are sanitized).

The things that make it hard for me to reason about are:
* the code that sanitizes is far away from the code that uses the headers
* the headers are sanitized in multiple places
* the sanitized headers are referred to by string literals, and there's no other hint about which headers are sanitized

Reply:
OK .. I will remove the sanitation from the finally block. In practice it should not matter unless there is a major refactoring of how/when this code is executed. 

> I meant the headers can all be referred by variables in SecurityRestFilter#handleRequest

Sorry if I am dense here .. but do you mean that the method should be threadContext.sanitizeHeaders(String... headersToSantize) , then call it from SecurityRestFilter#handleRequest with something like threadContext.sanitizeHeaders(CONSTANT_STRING, CONSTANT_STRING2) ? Since thread context lives in core I think that is the only way to make use variable names in SecurityRestFilter#handleRequest ?  However, if we do that then due to the public scope and global variable like nature we expose the ability for consumers to sanitize any header from anywhere and don't think we want that...but could be convinced otherwise since they can already stash from anywhere which isn't too dissimilar.

Reply:
>but do you mean that the method should be threadContext.sanitizeHeaders(String... headersToSantize) , then call it from SecurityRestFilter#handleRequest with something like threadContext.sanitizeHeaders(CONSTANT_STRING, CONSTANT_STRING2)

I was thinking that we can have the `sanitize` as a private method in the `SecurityRestFilter`, where the headers can be accessed by the variable name. I don't think that this is an issue since the sanitize method should never really be called anywhere else.

Reply:
We chatted a bit in person and are OK with it is as-is. Ideally we would use constants to refer to the header names and that constant would be used for all usages. In this case, the separation of core/x-pack makes that bit difficult since the primary usage is in x-pack but we need a reference in core. Moving the header to core is akward since it moves the definition from it's primary usage. Exposing which headers to sanitize (as mentioned above) isn't too desirable, and if were to sanitize only in the `SecurityRestFilter` it would require changing some private scope to public which is also not desirable. So we will just allow the duplication. These names are part of public documentation and so the names are pretty baked into our public API and changes to the names are pretty unlikely.